### PR TITLE
Allowlist verification - Enhancement #16

### PR DESCRIPTION
**Pull request by mpeters**
First draft at implementing Enhancement #16

* Add support for new JSON allowlist spec
  If the first non-whitespace character is a "{" then we treat it as
  JSON. Else we convert the old plain text format into the same structured
  data before passing it along
* Allow checksum support for the allowlist
* Allow GPG signature verification for the allowlist
* Allow pulling the allowlist and the signature from a URL

### CODE REVIEW

**Comment by mpeters** [comment]
Outstanding items:

* CLI parameter validation so things can only be combined in ways that make sense and requirements are validated.
* Better error handling
* Minor quibbles with the spec in the enhancement which we'll need to shake out

**Review by axelsimon — COMMENTED** [review]
Had a look, i don't see any glaring issues, but bear in mind my Python knowledge is still very limited. I'll defer to Luke and others to do a proper review.

File: keylime/ima.py [hunk 1]
```diff
@@ -24,6 +27,8 @@
 START_HASH = (codecs.decode('0000000000000000000000000000000000000000', 'hex'))
 FF_HASH = (codecs.decode('ffffffffffffffffffffffffffffffffffffffff', 'hex'))
 
+ALLOWLIST_CURRENT_VERSION = 1
```
Comment (ashcrow) [inline keylime/ima.py:30]: nit: Document constant

```suggestion
#: The version of the allowlist that is supported by this keylime release
ALLOWLIST_CURRENT_VERSION = 1
```

**Review by ashcrow — COMMENTED** [review]
Some nits but nothing that requires updating. Code looks great so far!

File: keylime/ima.py [hunk 2]
```diff
@@ -248,53 +253,106 @@ def process_measurement_list(lines, lists=None, m2w=None, pcrval=None):
...
-        allowlist[path] = tmp
-
-    if allowlist.get('boot_aggregate', None) is None:
-        logger.warning(
-            "No boot_aggregate value found in allowlist, adding an empty one")
-        allowlist['boot_aggregate'] = [
-            '0000000000000000000000000000000000000000']
-
-    for excl in excl_data:
+    if allowlist['hashes'].get('boot_aggregate', None) is None:
```
[the remark this answers is not in the document]
Comment (mpeters) [reply keylime/ima.py]: Done

File: keylime/ima.py [hunk 3]
```diff
@@ -248,53 +253,106 @@ def process_measurement_list(lines, lists=None, m2w=None, pcrval=None):
...
+        sha256.update(alist_bytes)
+        calculated_checksum = sha256.hexdigest()
+        alist_raw = alist_bytes.decode("utf-8")
+        logger.debug("Loaded allowlist from %s with checksum %s" % (al_path, calculated_checksum))
+
+    if checksum:
+        if checksum == calculated_checksum:
+            logger.debug("Allowlist passed checksum validation")
+        else:
+            raise Exception("Checksum of allowlist does not match! Expected %s, Calculated %s" % (checksum, calculated_checksum))
```
[the remark this answers is not in the document]
Comment (mpeters) [reply keylime/ima.py]: You're right, I'm still trying to train myself to use f-strings even after all this time :)

File: keylime/ima.py (hunk 1 above)
Comment (mpeters) [reply keylime/ima.py:31]: Done

**Comment by lukehinds** [comment]
Cursory look, shows this is looking good to me. Do we need to import a public key into a keyring for the verification part, I guess this might be a documentation thing?

**Comment by mpeters** [comment]
> Do we need to import a public key into a keyring for the verification part, I guess this might be a documentation thing?

No, it's setup that you provide the public key to keylime_tenant on the CLI.

**Comment by mpeters** [comment]
btw, this is still a WIP because I need to write tests for it. I keep trying to start that and it keeps being pushed off by other things. Will make a priority this week.

**Comment by mpeters** [comment]
Tests have been added and we're out of WIP, so ready for real evaluation now

**Comment by maugustosilva** [comment]
Tried to run the`verifier`, and got an error due to missing python module `gnupg`. Needs to be added to requirements.

**Comment by maugustosilva** [comment]
Maybe I am missing something, but wasn't it supposed to be backward compatible?

```
/usr/local/bin/keylime_tenant -c update -t 100.64.11.16 -v 100.64.11.15 -u 96A1B6FA-77B4-5F02-B3E2-0AEDE0DE617B --cert /var/lib/keylime/ca/96A1B6FA-77B4-5F02-B3E2-0AEDE0DE617B --mb_refstate /tmp/mb_refstate_96A1B6FA-77B4-5F02-B3E2-0AEDE0DE617B.txt --allowlist /tmp/whitelist_96A1B6FA-77B4-5F02-B3E2-0AEDE0DE617B.txt --exclude /ima_exclude.txt --tpm_policy '{"23":"0000000000000000000000000000000000000000000000000000000000000000"}'
Using config file /etc/keylime.conf
2021-03-11 17:40:08.435 - gnupg - INFO - Log opened: Thu Mar 11 17:40:08 2021 UTC
importing ['keylime-ibmextras.elchecking'], __package__='keylime.elchecking'
2021-03-11 17:40:08.498 - keylime.tpm - INFO - TPM2-TOOLS Version: 5.0
2021-03-11 17:40:08.502 - keylime.tenant - INFO - Setting up client TLS in /var/lib/keylime/cv_ca
2021-03-11 17:40:08.502 - keylime.tenant - INFO - TPM PCR Mask from policy is 0x800000
2021-03-11 17:40:08.502 - keylime.tenant - INFO - TPM PCR Mask from policy is 0x808000
2021-03-11 17:40:08.536 - keylime.ima - DEBUG - Loading allowlist from /tmp/whitelist_96A1B6FA-77B4-5F02-B3E2-0AEDE0DE617B.txt
2021-03-11 17:40:08.577 - keylime.ima - DEBUG - Loaded allowlist from /tmp/whitelist_96A1B6FA-77B4-5F02-B3E2-0AEDE0DE617B.txt with checksum 5a99b724c558ad1191e1dd0748161b71e919fd5e5fc36c646692657e4d7d1713
2021-03-11 17:40:08.577 - keylime.ima - DEBUG - Converting legacy allowlist format to JSON
Traceback (most recent call last):
  File "/usr/local/bin/keylime_tenant", line 11, in <module>
    load_entry_point('keylime==0.0.0', 'console_scripts', 'keylime_tenant')()
  File "/usr/local/lib/python3.8/site-packages/keylime-0.0.0-py3.8.egg/keylime/cmd/tenant.py", line 15, in main
    tenant.main()
  File "/usr/local/lib/python3.8/site-packages/keylime-0.0.0-py3.8.egg/keylime/tenant.py", line 1080, in main
    mytenant.init_add(vars(args))
  File "/usr/local/lib/python3.8/site-packages/keylime-0.0.0-py3.8.egg/keylime/tenant.py", line 223, in init_add
    al_data = ima.read_allowlist(args["allowlist"], args["allowlist_checksum"], args["allowlist_sig"], args["allowlist_sig_key"])
  File "/usr/local/lib/python3.8/site-packages/keylime-0.0.0-py3.8.egg/keylime/ima.py", line 469, in read_allowlist
    (checksum_hash, path) = line.split(None, 1)
ValueError: not enough values to unpack (expected 2, got 0)
```

where the contents of `allowlist` are (just the beginning)

```
sha256:4309217e9709def327c3409407f4d52907aeba46ef1b4f46529fd095f2ef9fb2  boot_aggregate

46286ac09df5d09d9b588f950837bb5e1a870404  /KL/keylime/.github/ISSUE_TEMPLATE.md
121d5c7fd8504219a2ba8ed70af80c9c54d26246  /KL/keylime/.github/workflows/greetings.yml
ae279cd3e82bcc74ab7076cf5e1ff4a0ca33884e  /KL/keylime/.github/workflows/semantic-release.yml
e8c1c8e7b55e987961eb5655ee31f1c729861f50  /KL/keylime/.github/workflows/keylime-bot.yml
5b14b8d133383cdfefd7cca751e32fe9a0550a71  /KL/keylime/.github/workflows/test.yml
f5df290c7c3fecd9ca445e22e38d540c8be2e7df  /KL/keylime/.github/stale.yml
1f621973dbe438207eb64c3efb2a77f2bccf3f8f  /KL/keylime/ROADMAP.md
06a388b72b39ca1ef558af373fc09f4adaef0525  /KL/keylime/CONTRIBUTORS.md
197226ca0a74ae6087de34bcc4d5d47f9f480825  /KL/keylime/archive/script_create_quote.py
22bef900d367a21b9eae8be860bd4c215ff3c0dc  /KL/keylime/archive/script_vtpm_register.sh
e50bb8210fb315826bd221a4b3166b58b10d80be  /KL/keylime/archive/script_old_quote.py
974708bd962f52b82da32b00bf69c342582d8e32  /KL/keylime/archive/script_check_quote.py
6451c6e45eda09f36516bbfe7d8061b33fe05da1  /KL/keylime/archive/ima/allowlist.txt
8e4ee61b9916e691a1c1c29a6e2406ef687b2efe  /KL/keylime/archive/ima/ascii_runtime_measurements_ima
b6d69bbc83f786fd14ad905b0674a63c682cfd35  /KL/keylime/archive/ima/exclude.txt
8f27cc51961b5facfd73da1b4512e7e2ac8ed616  /KL/keylime/archive/ima/ascii_runtime_measurements
8b7e3da906b0806df5115a036a2cd2c0acc664a0  /KL/keylime/archive/script_check_quote_raw.py
c1d3531f0b72eea3e575c09ee6876e9e70d4bb5e  /KL/keylime/archive/moc-new.bu.edu-pubek.pem
45fe2b7b94fc0bfc740f98dddcc2616fedabf917  /KL/keylime/archive/llsrc-vtpm-host0_pubek.pem
5043744ef4ef84c25da70394b9dc791501ded301  /KL/keylime/archive/README.md
...
```

**Comment by mpeters** [comment]
@maugustosilva it is supposed to be backwards compatible so it appears you found a bug :)

**Comment by maugustosilva** [comment]
> Tried to run the`verifier`, and got an error due to missing python module `gnupg`. Needs to be added to requirements.

@mpeters just to reiterate this point... I think the `requirements.txt` should list `gnupg` and not `python-gnupg` (certainly did not work for me on my local tests, I had to do `pip3 install gnupg` by hand

**Comment by mpeters** [comment]
> > Tried to run the`verifier`, and got an error due to missing python module `gnupg`. Needs to be added to requirements.
> 
> @mpeters just to reiterate this point... I think the `requirements.txt` should list `gnupg` and not `python-gnupg` (certainly did not work for me on my local tests, I had to do `pip3 install gnupg` by hand

It's actually really confusing because `gnupg` and `python-gnupg` are 2 separate packages, one forked from the other and they are not backwards compatible in subtle ways. And to add to the problem, you use both of them by saying `import gnupg` so depending on which one(s) are installed, you get something else. So it should be `python-gnupg` in requirements, but `gnupg` in import.

@maugustosilva  It's working for me and for CI, so I'm not sure why it's not working for you but we should track that down.

**Comment by mpeters** [comment]
> @maugustosilva it is supposed to be backwards compatible so it appears you found a bug :)

So my parser that converted the old format to a new format didn't handle blank lines well, this is now fixed in the latest version with the tests adjusted to cover that condition.

**Comment by mpeters** [comment]
> @maugustosilva it is supposed to be backwards compatible so it appears you found a bug :)

So my parser that converted the old format to a new format didn't handle blank lines well, this is now fixed in the latest version with the tests adjusted to cover that condition.

**Comment by maugustosilva** [comment]
I can confirm it is fully working for me now. LGTM

File: keylime/tenant.py [hunk 4]
```diff
@@ -1010,6 +1041,34 @@ def main(argv=sys.argv):
     if args.verifier_ip is not None:
         mytenant.verifier_ip = args.verifier_ip
 
+    # we only need to fetch remote files if we are adding or updateing
+    if args.command in ['add', 'update']:
+        delete_tmp_files = logger.level > logging.DEBUG # delete tmp files unless in DEBUG mode
+
+        if args.allowlist_url:
+            logger.info("Downloading Allowlist from %s" % (args.allowlist_url))
```
Comment (ashcrow) [inline keylime/tenant.py:1049]: nit: Consider using `logging`s string templating directly here and in other log statements.

```suggestion
            logger.info("Downloading Allowlist from %s", args.allowlist_url)
```

**Review by ashcrow — APPROVED** [review]

File: keylime/tenant.py (hunk 4 above)
Comment (mpeters) [reply keylime/tenant.py:1049]: Agreed and fixed. This is something we also need for the rest of keylime, but I'll do that in another PR

**Review by ashcrow — APPROVED** [review]

**Review by lukehinds — APPROVED** [review]

**Pull request merged**